File: tests/place-bet.ts
import * as anchor from "@coral-xyz/anchor";
import { Program } from "@coral-xyz/anchor";
import {
  PublicKey,
  Keypair,
  SystemProgram,
  LAMPORTS_PER_SOL,
} from "@solana/web3.js";
import {
  createMint,
  createAssociatedTokenAccount,
  mintTo,
  getAccount,
  TOKEN_PROGRAM_ID,
} from "@solana/spl-token";
import { Avenir } from "../target/types/avenir";
import { assert } from "chai";
import { randomBytes } from "crypto";

import {
  setupArciumContext,
  encryptBetInput,
  createTestMarket,
  getArciumAccounts,
  initCompDef,
  getConfigPda,
  getMarketPda,
  getMarketPoolPda,
  ArciumContext,
} from "./mpc/helpers";

describe("place_bet", () => {
  const provider = anchor.AnchorProvider.env();
  anchor.setProvider(provider);

  const program = anchor.workspace.avenir as Program<Avenir>;
  const admin = provider.wallet as anchor.Wallet;
  const payer = admin.payer;
  const connection = provider.connection;

  // Test keypairs
  const creator = Keypair.generate();
  const bettor = Keypair.generate();
  const bettor2 = Keypair.generate();
  const mintAuthority = Keypair.generate();
  const feeRecipient = Keypair.generate().publicKey;
  const protocolFeeBps = 200; // 2%

  // Will be set in before() hooks
  let usdcMint: PublicKey;
  let arciumCtx: ArciumContext;
  let bettorTokenAccount: PublicKey;
  let bettor2TokenAccount: PublicKey;

  // Config PDA
  const [configPda] = getConfigPda(program.programId);

  // Market state (set in before hook)
  let marketPda: PublicKey;
  let marketPoolPda: PublicKey;
  let vaultPda: PublicKey;
  let marketId: number;

  // PDA helpers
  function getPositionPda(
    mktId: number,
    user: PublicKey
  ): [PublicKey, number] {
    const buf = Buffer.alloc(8);
    buf.writeBigUInt64LE(BigInt(mktId));
    return PublicKey.findProgramAddressSync(
      [Buffer.from("position"), buf, user.toBuffer()],
      program.programId
    );
  }

  function getVaultPda(mktId: number): [PublicKey, number] {
    const buf = Buffer.alloc(8);
    buf.writeBigUInt64LE(BigInt(mktId));
    return PublicKey.findProgramAddressSync(
      [Buffer.from("vault"), buf],
      program.programId
    );
  }

  before(async () => {
    // 1. Airdrop SOL to test accounts
    const airdropCreator = await connection.requestAirdrop(
      creator.publicKey,
      10 * LAMPORTS_PER_SOL
    );
    await connection.confirmTransaction(airdropCreator);

    const airdropBettor = await connection.requestAirdrop(
      bettor.publicKey,
      10 * LAMPORTS_PER_SOL
    );
    await connection.confirmTransaction(airdropBettor);

    const airdropBettor2 = await connection.requestAirdrop(
      bettor2.publicKey,
      10 * LAMPORTS_PER_SOL
    );
    await connection.confirmTransaction(airdropBettor2);

    const airdropMintAuth = await connection.requestAirdrop(
      mintAuthority.publicKey,
      2 * LAMPORTS_PER_SOL
    );
    await connection.confirmTransaction(airdropMintAuth);

    // 2. Create USDC mint (6 decimals)
    usdcMint = await createMint(
      connection,
      mintAuthority,
      mintAuthority.publicKey,
      null,
      6
    );

    // 3. Initialize Config PDA
    await program.methods
      .initialize({
        feeRecipient,
        usdcMint,
        protocolFeeBps,
      })
      .accountsPartial({
        admin: admin.publicKey,
        config: configPda,
        systemProgram: SystemProgram.programId,
      })
      .rpc({ commitment: "confirmed" });

    // 4. Whitelist the test creator
    const [whitelistPda] = PublicKey.findProgramAddressSync(
      [Buffer.from("whitelist"), creator.publicKey.toBuffer()],
      program.programId
    );

    await program.methods
      .addCreator()
      .accountsPartial({
        admin: admin.publicKey,
        config: configPda,
        whitelist: whitelistPda,
        creator: creator.publicKey,
        systemProgram: SystemProgram.programId,
      })
      .rpc({ commitment: "confirmed" });

    // 5. Setup Arcium context
    arciumCtx = await setupArciumContext(provider, program.programId);

    // 6. Initialize comp_defs for update_pool (needed by place_bet)
    await initCompDef(program, payer, "init_pool", arciumCtx);
    await initCompDef(program, payer, "update_pool", arciumCtx);

    // 7. Create a market with future deadline
    const result = await createTestMarket(
      program,
      admin,
      creator,
      usdcMint,
      configPda,
      1, // expectedMarketId
      {
        question: "Will place_bet tests pass?",
        resolutionTime: Math.floor(Date.now() / 1000) + 7200, // 2h from now
      }
    );
    marketPda = result.marketPda;
    marketPoolPda = result.marketPoolPda;
    vaultPda = result.vaultPda;
    marketId = result.marketId;

    // 8. Initialize pool via init_pool MPC to get valid encrypted zeros
    const initPoolBet = encryptBetInput(arciumCtx.mxePublicKey, true, 0);
    const initPoolAccounts = getArciumAccounts(
      program.programId,
      arciumCtx,
      initPoolBet.computationOffset,
      "init_pool"
    );

    await program.methods
      .initPool(initPoolBet.computationOffset)
      .accountsPartial({
        payer: payer.publicKey,
        marketPool: marketPoolPda,
        mxeAccount: initPoolAccounts.mxeAccount,
        signPdaAccount: initPoolAccounts.signPdaAccount,
        mempoolAccount: initPoolAccounts.mempoolAccount,
        executingPool: initPoolAccounts.executingPool,
        computationAccount: initPoolAccounts.computationAccount,
        compDefAccount: initPoolAccounts.compDefAccount,
        clusterAccount: initPoolAccounts.clusterAccount,
        poolAccount: initPoolAccounts.poolAccount,
        clockAccount: initPoolAccounts.clockAccount,
        systemProgram: SystemProgram.programId,
        arciumProgram: initPoolAccounts.arciumProgram,
      })
      .rpc({ skipPreflight: true, commitment: "confirmed" });

    // Wait for init_pool callback to complete
    const { awaitComputationFinalization } = await import("@arcium-hq/client");
    await awaitComputationFinalization(
      provider,
      initPoolBet.computationOffset,
      program.programId,
      "confirmed",
      120_000
    );

    // 9. Create bettor's USDC token accounts and mint test USDC
    bettorTokenAccount = await createAssociatedTokenAccount(
      connection,
      bettor,
      usdcMint,
      bettor.publicKey
    );

    // Mint 100 USDC to bettor
    await mintTo(
      connection,
      mintAuthority,
      usdcMint,
      bettorTokenAccount,
      mintAuthority.publicKey,
      100_000_000 // 100 USDC
    );

    bettor2TokenAccount = await createAssociatedTokenAccount(
      connection,
      bettor2,
      usdcMint,
      bettor2.publicKey
    );

    // Mint 100 USDC to bettor2
    await mintTo(
      connection,
      mintAuthority,
      usdcMint,
      bettor2TokenAccount,
      mintAuthority.publicKey,
      100_000_000 // 100 USDC
    );

    console.log("  place_bet test setup complete:");
    console.log("    USDC Mint:", usdcMint.toBase58());
    console.log("    Market ID:", marketId);
    console.log("    Bettor:", bettor.publicKey.toBase58());
    console.log("    Bettor2:", bettor2.publicKey.toBase58());
  });

  // ==========================================================================
  // Test 1: places a valid Yes bet with USDC transfer
  // ==========================================================================
  it("places a valid Yes bet with USDC transfer", async () => {
    const betAmount = 5_000_000; // 5 USDC
    const isYes = true;

    // Get balances before
    const bettorBalanceBefore = (await getAccount(connection, bettorTokenAccount)).amount;
    const vaultBalanceBefore = (await getAccount(connection, vaultPda)).amount;

    // Encrypt bet input
    const bet = encryptBetInput(arciumCtx.mxePublicKey, isYes, betAmount);

    const arciumAccounts = getArciumAccounts(
      program.programId,
      arciumCtx,
      bet.computationOffset,
      "update_pool"
    );

    // Derive position PDA
    const [positionPda] = getPositionPda(marketId, bettor.publicKey);

    await program.methods
      .placeBet(
        new anchor.BN(betAmount),
        isYes,
        new anchor.BN(bet.computationOffset),
        Array.from(bet.isYesCiphertext) as unknown as number[],
        Array.from(bet.amountCiphertext) as unknown as number[],
        Array.from(bet.publicKey) as unknown as number[],
        bet.nonceBN
      )
      .accountsPartial({
        bettor: bettor.publicKey,
        market: marketPda,
        marketPool: marketPoolPda,
        userPosition: positionPda,
        userTokenAccount: bettorTokenAccount,
        marketVault: vaultPda,
        pendingBettorTokenAccount: bettorTokenAccount,
        usdcMint: usdcMint,
        tokenProgram: TOKEN_PROGRAM_ID,
        systemProgram: SystemProgram.programId,
        mxeAccount: arciumAccounts.mxeAccount,
        signPdaAccount: arciumAccounts.signPdaAccount,
        mempoolAccount: arciumAccounts.mempoolAccount,
        executingPool: arciumAccounts.executingPool,
        computationAccount: arciumAccounts.computationAccount,
        compDefAccount: arciumAccounts.compDefAccount,
        clusterAccount: arciumAccounts.clusterAccount,
        poolAccount: arciumAccounts.poolAccount,
        clockAccount: arciumAccounts.clockAccount,
        arciumProgram: arciumAccounts.arciumProgram,
      })
      .signers([bettor])
      .rpc({ skipPreflight: true, commitment: "confirmed" });

    // Verify: bettor's token balance decreased by 5 USDC
    const bettorBalanceAfter = (await getAccount(connection, bettorTokenAccount)).amount;
    assert.equal(
      Number(bettorBalanceBefore - bettorBalanceAfter),
      betAmount,
      "Bettor balance should decrease by bet amount"
    );

    // Verify: vault balance increased by 5 USDC
    const vaultBalanceAfter = (await getAccount(connection, vaultPda)).amount;
    assert.equal(
      Number(vaultBalanceAfter - vaultBalanceBefore),
      betAmount,
      "Vault balance should increase by bet amount"
    );

    // Verify: market state
    const market = await program.account.market.fetch(marketPda);
    assert.isTrue(market.mpcLock, "mpc_lock should be true");
    assert.ok(
      market.pendingBettor.equals(bettor.publicKey),
      "pending_bettor should be bettor"
    );
    assert.equal(
      market.pendingAmount.toNumber(),
      betAmount,
      "pending_amount should be bet amount"
    );
    assert.isTrue(market.pendingIsYes, "pending_is_yes should be true");
    assert.isTrue(
      market.lockTimestamp.toNumber() > 0,
      "lock_timestamp should be set"
    );

    // Verify: UserPosition PDA exists (created by init_if_needed)
    const position = await program.account.userPosition.fetch(positionPda);
    assert.equal(
      position.marketId.toNumber(),
      marketId,
      "UserPosition market_id should match"
    );

    console.log("    Valid Yes bet placed:");
    console.log("      Amount:", betAmount, "lamports (5 USDC)");
    console.log("      Bettor balance change:", Number(bettorBalanceBefore - bettorBalanceAfter));
    console.log("      Vault balance change:", Number(vaultBalanceAfter - vaultBalanceBefore));
    console.log("      mpc_lock:", market.mpcLock);
    console.log("      lock_timestamp:", market.lockTimestamp.toNumber());

    // Wait for MPC callback to complete so lock is released for subsequent tests
    console.log("    Awaiting MPC callback to release lock...");
    const { awaitComputationFinalization } = await import("@arcium-hq/client");
    await awaitComputationFinalization(
      provider,
      bet.computationOffset,
      program.programId,
      "confirmed",
      120_000
    );

    // Verify lock released
    const marketAfterCallback = await program.account.market.fetch(marketPda);
    assert.isFalse(marketAfterCallback.mpcLock, "mpc_lock should be released after callback");
    console.log("    MPC callback complete, lock released");
  });

  // ==========================================================================
  // Test 2: rejects bet below minimum (1 USDC)
  // ==========================================================================
  it("rejects bet below minimum (1 USDC)", async () => {
    const betAmount = 999_999; // Just under 1 USDC
    const isYes = true;

    const bet = encryptBetInput(arciumCtx.mxePublicKey, isYes, betAmount);
    const arciumAccounts = getArciumAccounts(
      program.programId,
      arciumCtx,
      bet.computationOffset,
      "update_pool"
    );
    const [positionPda] = getPositionPda(marketId, bettor.publicKey);

    // Record balance before
    const bettorBalanceBefore = (await getAccount(connection, bettorTokenAccount)).amount;

    try {
      await program.methods
        .placeBet(
          new anchor.BN(betAmount),
          isYes,
          new anchor.BN(bet.computationOffset),
          Array.from(bet.isYesCiphertext) as unknown as number[],
          Array.from(bet.amountCiphertext) as unknown as number[],
          Array.from(bet.publicKey) as unknown as number[],
          bet.nonceBN
        )
        .accountsPartial({
          bettor: bettor.publicKey,
          market: marketPda,
          marketPool: marketPoolPda,
          userPosition: positionPda,
          userTokenAccount: bettorTokenAccount,
          marketVault: vaultPda,
          pendingBettorTokenAccount: bettorTokenAccount,
          usdcMint: usdcMint,
          tokenProgram: TOKEN_PROGRAM_ID,
          systemProgram: SystemProgram.programId,
          mxeAccount: arciumAccounts.mxeAccount,
          signPdaAccount: arciumAccounts.signPdaAccount,
          mempoolAccount: arciumAccounts.mempoolAccount,
          executingPool: arciumAccounts.executingPool,
          computationAccount: arciumAccounts.computationAccount,
          compDefAccount: arciumAccounts.compDefAccount,
          clusterAccount: arciumAccounts.clusterAccount,
          poolAccount: arciumAccounts.poolAccount,
          clockAccount: arciumAccounts.clockAccount,
          arciumProgram: arciumAccounts.arciumProgram,
        })
        .signers([bettor])
        .rpc({ skipPreflight: true, commitment: "confirmed" });

      assert.fail("Expected BetTooSmall error");
    } catch (e: any) {
      const errStr = e.toString();
      assert.isTrue(
        errStr.includes("BetTooSmall") ||
          errStr.includes("6012") ||
          errStr.includes("Bet amount must be at least 1 USDC"),
        `Expected BetTooSmall error, got: ${errStr.substring(0, 200)}`
      );
      console.log("    BetTooSmall correctly rejected");
    }

    // Verify no USDC transferred
    const bettorBalanceAfter = (await getAccount(connection, bettorTokenAccount)).amount;
    assert.equal(
      Number(bettorBalanceBefore),
      Number(bettorBalanceAfter),
      "Bettor balance should be unchanged after rejection"
    );
  });

  // ==========================================================================
  // Test 3: rejects bet when MPC lock is active
  // ==========================================================================
  it("rejects bet when MPC lock is active", async () => {
    // Place a valid bet to activate the lock
    const bet1Amount = 1_000_000; // 1 USDC
    const bet1 = encryptBetInput(arciumCtx.mxePublicKey, true, bet1Amount);
    const arciumAccounts1 = getArciumAccounts(
      program.programId,
      arciumCtx,
      bet1.computationOffset,
      "update_pool"
    );
    const [positionPda1] = getPositionPda(marketId, bettor.publicKey);

    await program.methods
      .placeBet(
        new anchor.BN(bet1Amount),
        true,
        new anchor.BN(bet1.computationOffset),
        Array.from(bet1.isYesCiphertext) as unknown as number[],
        Array.from(bet1.amountCiphertext) as unknown as number[],
        Array.from(bet1.publicKey) as unknown as number[],
        bet1.nonceBN
      )
      .accountsPartial({
        bettor: bettor.publicKey,
        market: marketPda,
        marketPool: marketPoolPda,
        userPosition: positionPda1,
        userTokenAccount: bettorTokenAccount,
        marketVault: vaultPda,
        pendingBettorTokenAccount: bettorTokenAccount,
        usdcMint: usdcMint,
        tokenProgram: TOKEN_PROGRAM_ID,
        systemProgram: SystemProgram.programId,
        mxeAccount: arciumAccounts1.mxeAccount,
        signPdaAccount: arciumAccounts1.signPdaAccount,
        mempoolAccount: arciumAccounts1.mempoolAccount,
        executingPool: arciumAccounts1.executingPool,
        computationAccount: arciumAccounts1.computationAccount,
        compDefAccount: arciumAccounts1.compDefAccount,
        clusterAccount: arciumAccounts1.clusterAccount,
        poolAccount: arciumAccounts1.poolAccount,
        clockAccount: arciumAccounts1.clockAccount,
        arciumProgram: arciumAccounts1.arciumProgram,
      })
      .signers([bettor])
      .rpc({ skipPreflight: true, commitment: "confirmed" });

    console.log("    First bet placed, mpc_lock should be active");

    // Verify lock is active
    const marketLocked = await program.account.market.fetch(marketPda);
    assert.isTrue(marketLocked.mpcLock, "mpc_lock should be true");

    // Attempt second bet from bettor2 while lock is active
    const bet2Amount = 2_000_000; // 2 USDC
    const bet2 = encryptBetInput(arciumCtx.mxePublicKey, true, bet2Amount);
    const arciumAccounts2 = getArciumAccounts(
      program.programId,
      arciumCtx,
      bet2.computationOffset,
      "update_pool"
    );
    const [positionPda2] = getPositionPda(marketId, bettor2.publicKey);

    // Record bettor2 balance before
    const bettor2BalanceBefore = (await getAccount(connection, bettor2TokenAccount)).amount;

    try {
      await program.methods
        .placeBet(
          new anchor.BN(bet2Amount),
          true,
          new anchor.BN(bet2.computationOffset),
          Array.from(bet2.isYesCiphertext) as unknown as number[],
          Array.from(bet2.amountCiphertext) as unknown as number[],
          Array.from(bet2.publicKey) as unknown as number[],
          bet2.nonceBN
        )
        .accountsPartial({
          bettor: bettor2.publicKey,
          market: marketPda,
          marketPool: marketPoolPda,
          userPosition: positionPda2,
          userTokenAccount: bettor2TokenAccount,
          marketVault: vaultPda,
          pendingBettorTokenAccount: bettor2TokenAccount,
          usdcMint: usdcMint,
          tokenProgram: TOKEN_PROGRAM_ID,
          systemProgram: SystemProgram.programId,
          mxeAccount: arciumAccounts2.mxeAccount,
          signPdaAccount: arciumAccounts2.signPdaAccount,
          mempoolAccount: arciumAccounts2.mempoolAccount,
          executingPool: arciumAccounts2.executingPool,
          computationAccount: arciumAccounts2.computationAccount,
          compDefAccount: arciumAccounts2.compDefAccount,
          clusterAccount: arciumAccounts2.clusterAccount,
          poolAccount: arciumAccounts2.poolAccount,
          clockAccount: arciumAccounts2.clockAccount,
          arciumProgram: arciumAccounts2.arciumProgram,
        })
        .signers([bettor2])
        .rpc({ skipPreflight: true, commitment: "confirmed" });

      assert.fail("Expected MpcLocked error");
    } catch (e: any) {
      const errStr = e.toString();
      assert.isTrue(
        errStr.includes("MpcLocked") ||
          errStr.includes("6012") ||
          errStr.includes("Market MPC computation is in progress"),
        `Expected MpcLocked error, got: ${errStr.substring(0, 200)}`
      );
      console.log("    MpcLocked correctly rejected for bettor2");
    }

    // Verify bettor2 balance unchanged
    const bettor2BalanceAfter = (await getAccount(connection, bettor2TokenAccount)).amount;
    assert.equal(
      Number(bettor2BalanceBefore),
      Number(bettor2BalanceAfter),
      "Bettor2 balance should be unchanged after MpcLocked rejection"
    );

    // Wait for the first bet's MPC callback to release the lock
    console.log("    Awaiting MPC callback to release lock...");
    const { awaitComputationFinalization } = await import("@arcium-hq/client");
    await awaitComputationFinalization(
      provider,
      bet1.computationOffset,
      program.programId,
      "confirmed",
      120_000
    );

    const marketUnlocked = await program.account.market.fetch(marketPda);
    assert.isFalse(marketUnlocked.mpcLock, "mpc_lock should be released after callback");
    console.log("    Lock released after callback");
  });

  // ==========================================================================
  // Test 4: validates UserPosition PDA creation via init_if_needed
  // ==========================================================================
  it("validates UserPosition PDA creation via init_if_needed", async () => {
    // After test 1, the UserPosition PDA should exist for bettor on this market
    const [positionPda] = getPositionPda(marketId, bettor.publicKey);

    const position = await program.account.userPosition.fetch(positionPda);

    // Verify: account exists with correct market_id
    assert.equal(
      position.marketId.toNumber(),
      marketId,
      "UserPosition market_id should match"
    );

    // Verify: user is set to bettor (set by init_if_needed in place_bet)
    assert.ok(
      position.user.equals(bettor.publicKey),
      "UserPosition user should be the bettor"
    );

    // After callback has run (test 1 waited for it), yes_amount should be accumulated
    // The bettor placed 5 USDC (test 1) + 1 USDC (test 3) = 6 USDC on Yes
    assert.isTrue(
      position.yesAmount.toNumber() > 0,
      "UserPosition yes_amount should be > 0 after callback"
    );

    // No amount should be 0 (bettor only bet Yes)
    assert.equal(
      position.noAmount.toNumber(),
      0,
      "UserPosition no_amount should be 0"
    );

    // claimed should be false
    assert.isFalse(position.claimed, "UserPosition claimed should be false");

    console.log("    UserPosition PDA verified:");
    console.log("      market_id:", position.marketId.toNumber());
    console.log("      user:", position.user.toBase58());
    console.log("      yes_amount:", position.yesAmount.toNumber());
    console.log("      no_amount:", position.noAmount.toNumber());
    console.log("      claimed:", position.claimed);
  });

  // ==========================================================================
  // Test 5: rejects bet on expired market
  // ==========================================================================
  it("rejects bet on expired market", async () => {
    // Cannot create an expired market on localnet because create_market validates
    // deadline > now + 1h. Clock sysvar manipulation is not available on localnet.
    //
    // Instead, verify the MarketExpired error variant exists in the IDL,
    // confirming the validation guard is compiled into the program.
    // The actual require! check:
    //   require!(clock.unix_timestamp < market.resolution_time, AvenirError::MarketExpired);
    // is verified by code inspection and the fact that the program compiles.

    const idlErrors = program.idl.errors;
    assert.isDefined(idlErrors, "Program IDL should have errors defined");

    const marketExpiredError = idlErrors!.find(
      (e: any) => e.name === "marketExpired" || e.name === "MarketExpired"
    );
    assert.isDefined(
      marketExpiredError,
      "MarketExpired error variant should exist in the program IDL"
    );

    console.log("    MarketExpired IDL error verified:");
    console.log("      name:", marketExpiredError!.name);
    console.log("      code:", marketExpiredError!.code);
    console.log("      msg:", marketExpiredError!.msg);
    console.log("    Note: Runtime test skipped (requires Clock sysvar manipulation)");
  });

  // ==========================================================================
  // Test 6: rejects bet on opposite side of existing position (WrongSide)
  // ==========================================================================
  it("rejects bet on opposite side of existing position", async () => {
    // After tests 1 and 3, bettor has a Yes position on this market.
    // Attempting a No bet should be rejected with WrongSide.

    const betAmount = 1_000_000; // 1 USDC
    const isYes = false; // Opposite side

    const bet = encryptBetInput(arciumCtx.mxePublicKey, isYes, betAmount);
    const arciumAccounts = getArciumAccounts(
      program.programId,
      arciumCtx,
      bet.computationOffset,
      "update_pool"
    );
    const [positionPda] = getPositionPda(marketId, bettor.publicKey);

    try {
      await program.methods
        .placeBet(
          new anchor.BN(betAmount),
          isYes,
          new anchor.BN(bet.computationOffset),
          Array.from(bet.isYesCiphertext) as unknown as number[],
          Array.from(bet.amountCiphertext) as unknown as number[],
          Array.from(bet.publicKey) as unknown as number[],
          bet.nonceBN
        )
        .accountsPartial({
          bettor: bettor.publicKey,
          market: marketPda,
          marketPool: marketPoolPda,
          userPosition: positionPda,
          userTokenAccount: bettorTokenAccount,
          marketVault: vaultPda,
          pendingBettorTokenAccount: bettorTokenAccount,
          usdcMint: usdcMint,
          tokenProgram: TOKEN_PROGRAM_ID,
          systemProgram: SystemProgram.programId,
          mxeAccount: arciumAccounts.mxeAccount,
          signPdaAccount: arciumAccounts.signPdaAccount,
          mempoolAccount: arciumAccounts.mempoolAccount,
          executingPool: arciumAccounts.executingPool,
          computationAccount: arciumAccounts.computationAccount,
          compDefAccount: arciumAccounts.compDefAccount,
          clusterAccount: arciumAccounts.clusterAccount,
          poolAccount: arciumAccounts.poolAccount,
          clockAccount: arciumAccounts.clockAccount,
          arciumProgram: arciumAccounts.arciumProgram,
        })
        .signers([bettor])
        .rpc({ skipPreflight: true, commitment: "confirmed" });

      assert.fail("Expected WrongSide error");
    } catch (e: any) {
      const errStr = e.toString();
      assert.isTrue(
        errStr.includes("WrongSide") ||
          errStr.includes("6014") ||
          errStr.includes("Cannot bet on opposite side"),
        `Expected WrongSide error, got: ${errStr.substring(0, 200)}`
      );
      console.log("    WrongSide correctly rejected (bettor has Yes position, attempted No bet)");
    }
  });
});

File: programs/avenir/src/instructions/place_bet.rs
use anchor_lang::prelude::*;
use anchor_spl::token::{self, Mint, Token, TokenAccount, Transfer};
use arcium_anchor::prelude::*;
use arcium_client::idl::arcium::types::CallbackAccount;

use crate::errors::{AvenirError, ErrorCode};
use crate::state::{Market, MarketPool, UserPosition};
use crate::{ArciumSignerAccount, ID, ID_CONST};

use super::mpc::update_pool_callback::UpdatePoolCallback;

/// Place a bet on a prediction market.
///
/// This instruction atomically:
/// 1. Validates bet amount (>= 1 USDC), market state (Open), deadline, and lock status
/// 2. Handles stale lock timeout recovery (refunds stuck bettor if lock > 60s old)
/// 3. Validates side consistency (cannot bet on opposite side of existing position)
/// 4. Transfers USDC from user to market vault
/// 5. Pre-creates UserPosition PDA via init_if_needed (Arcium callbacks cannot create accounts)
/// 6. Stores pending bet data on Market for callback to reference
/// 7. Queues the update_pool MPC computation with extended callback accounts
pub fn handler(
    ctx: Context<PlaceBet>,
    amount: u64,
    is_yes: bool,
    computation_offset: u64,
    is_yes_ciphertext: [u8; 32],
    amount_ciphertext: [u8; 32],
    pub_key: [u8; 32],
    nonce: u128,
) -> Result<()> {
    // 1. Get clock
    let clock = Clock::get()?;

    // 2. Validate bet amount (minimum 1 USDC = 1,000,000 token units)
    require!(amount >= 1_000_000, AvenirError::BetTooSmall);

    let market = &mut ctx.accounts.market;

    // 3. Validate market state
    require!(market.state == 0, AvenirError::MarketNotOpen);

    // 4. Validate deadline
    require!(
        clock.unix_timestamp < market.resolution_time,
        AvenirError::MarketExpired
    );

    // 5. Check lock with timeout recovery
    if market.mpc_lock {
        if clock.unix_timestamp - market.lock_timestamp > 60 {
            // Lock is stale (>60s) -- refund the stuck bettor
            let market_id = market.id;
            let bump = market.bump;
            let signer_seeds: &[&[&[u8]]] =
                &[&[b"market", &market_id.to_le_bytes(), &[bump]]];

            // Deserialize pending_bettor_token_account as TokenAccount for refund
            let pending_token_info = ctx.accounts.pending_bettor_token_account.to_account_info();

            let refund_accounts = Transfer {
                from: ctx.accounts.market_vault.to_account_info(),
                to: pending_token_info,
                authority: market.to_account_info(),
            };
            token::transfer(
                CpiContext::new_with_signer(
                    ctx.accounts.token_program.to_account_info(),
                    refund_accounts,
                    signer_seeds,
                ),
                market.pending_amount,
            )?;

            // Clear lock state
            market.mpc_lock = false;
            market.pending_bettor = Pubkey::default();
            market.pending_amount = 0;
            market.pending_is_yes = false;
            market.lock_timestamp = 0;
        } else {
            // Lock is active -- reject immediately
            return Err(AvenirError::MpcLocked.into());
        }
    }

    // 6. Validate side consistency
    let user_position = &ctx.accounts.user_position;
    if user_position.yes_amount > 0 && !is_yes {
        return Err(AvenirError::WrongSide.into());
    }
    if user_position.no_amount > 0 && is_yes {
        return Err(AvenirError::WrongSide.into());
    }

    // 7. Transfer USDC: user -> vault
    let cpi_accounts = Transfer {
        from: ctx.accounts.user_token_account.to_account_info(),
        to: ctx.accounts.market_vault.to_account_info(),
        authority: ctx.accounts.bettor.to_account_info(),
    };
    token::transfer(
        CpiContext::new(
            ctx.accounts.token_program.to_account_info(),
            cpi_accounts,
        ),
        amount,
    )?;

    // 8. Store pending bet data on Market
    market.pending_bettor = ctx.accounts.bettor.key();
    market.pending_amount = amount;
    market.pending_is_yes = is_yes;
    market.mpc_lock = true;
    market.lock_timestamp = clock.unix_timestamp;

    // 9. Build ArgBuilder and queue MPC computation
    let args = ArgBuilder::new()
        .x25519_pubkey(pub_key)
        .plaintext_u128(nonce)
        .encrypted_bool(is_yes_ciphertext)
        .encrypted_u64(amount_ciphertext)
        .account(
            ctx.accounts.market_pool.key(),
            16, // offset: 8 (discriminator) + 8 (market_id)
            64, // length: 32 (yes_pool) + 32 (no_pool)
        )
        .build();

    // 10. Build extended CallbackAccount vector
    let callback_accounts = vec![
        CallbackAccount {
            pubkey: ctx.accounts.market_pool.key(),
            is_writable: true,
        },
        CallbackAccount {
            pubkey: ctx.accounts.market.key(),
            is_writable: true,
        },
        CallbackAccount {
            pubkey: ctx.accounts.user_position.key(),
            is_writable: true,
        },
        CallbackAccount {
            pubkey: ctx.accounts.market_vault.key(),
            is_writable: true,
        },
        CallbackAccount {
            pubkey: ctx.accounts.user_token_account.key(),
            is_writable: true,
        },
        CallbackAccount {
            pubkey: ctx.accounts.token_program.key(),
            is_writable: false,
        },
    ];

    // 11. Build callback instruction and queue computation
    let callback_ixs = vec![UpdatePoolCallback::callback_ix(
        computation_offset,
        &ctx.accounts.mxe_account,
        &callback_accounts,
    )?];

    queue_computation(
        ctx.accounts,
        computation_offset,
        args,
        callback_ixs,
        1,
        0,
    )?;

    Ok(())
}

#[queue_computation_accounts("update_pool", bettor)]
#[derive(Accounts)]
pub struct PlaceBet<'info> {
    #[account(mut)]
    pub bettor: Signer<'info>,

    /// The Market account -- validates state, deadline, lock, stores pending bet data.
    #[account(
        mut,
        seeds = [b"market", market.id.to_le_bytes().as_ref()],
        bump = market.bump,
    )]
    pub market: Account<'info, Market>,

    /// The MarketPool PDA containing encrypted pool state.
    #[account(
        seeds = [b"market_pool", market_pool.market_id.to_le_bytes().as_ref()],
        bump = market_pool.bump,
        constraint = market_pool.market_id == market.id,
    )]
    pub market_pool: Account<'info, MarketPool>,

    /// Per-user, per-market position tracking.
    /// Created via init_if_needed because Arcium callbacks cannot create accounts.
    #[account(
        init_if_needed,
        payer = bettor,
        space = 8 + UserPosition::INIT_SPACE,
        seeds = [b"position", market.id.to_le_bytes().as_ref(), bettor.key().as_ref()],
        bump,
    )]
    pub user_position: Account<'info, UserPosition>,

    /// Bettor's USDC token account (source for bet transfer).
    #[account(
        mut,
        token::mint = usdc_mint,
        token::authority = bettor,
    )]
    pub user_token_account: Account<'info, TokenAccount>,

    /// Market vault token account (destination for bet transfer).
    #[account(
        mut,
        seeds = [b"vault", market.id.to_le_bytes().as_ref()],
        bump = market.vault_bump,
        token::mint = usdc_mint,
        token::authority = market,
    )]
    pub market_vault: Account<'info, TokenAccount>,

    /// CHECK: Only used for timeout refund -- validated against market.pending_bettor
    /// when timeout recovery triggers. When no timeout, this account is unused.
    #[account(mut)]
    pub pending_bettor_token_account: UncheckedAccount<'info>,

    /// USDC mint for token account constraints.
    pub usdc_mint: Account<'info, Mint>,

    pub token_program: Program<'info, Token>,
    pub system_program: Program<'info, System>,

    // Arcium MPC accounts (injected by queue_computation_accounts macro)
    #[account(
        address = derive_mxe_pda!()
    )]
    pub mxe_account: Account<'info, MXEAccount>,

    #[account(
        mut,
        address = derive_sign_pda!()
    )]
    pub sign_pda_account: Account<'info, ArciumSignerAccount>,

    #[account(
        mut,
        address = derive_mempool_pda!(mxe_account, ErrorCode::ClusterNotSet)
    )]
    /// CHECK: mempool_account, derived from cluster
    pub mempool_account: UncheckedAccount<'info>,

    #[account(
        mut,
        address = derive_execpool_pda!(mxe_account, ErrorCode::ClusterNotSet)
    )]
    /// CHECK: executing_pool, derived from cluster
    pub executing_pool: UncheckedAccount<'info>,

    #[account(mut)]
    /// CHECK: computation_account, initialized by arcium
    pub computation_account: UncheckedAccount<'info>,

    #[account(
        address = derive_comp_def_pda!(comp_def_offset("update_pool"))
    )]
    pub comp_def_account: Account<'info, ComputationDefinitionAccount>,

    #[account(
        mut,
        address = derive_cluster_pda!(mxe_account, ErrorCode::ClusterNotSet)
    )]
    pub cluster_account: Account<'info, Cluster>,

    #[account(
        mut,
        address = ARCIUM_FEE_POOL_ACCOUNT_ADDRESS,
    )]
    pub pool_account: Account<'info, FeePool>,

    #[account(
        mut,
        address = ARCIUM_CLOCK_ACCOUNT_ADDRESS
    )]
    pub clock_account: Account<'info, ClockAccount>,

    pub arcium_program: Program<'info, Arcium>,
}
